src/plugins/content_management/server/plugin.test.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { loggingSystemMock, coreMock } from '@kbn/core/server/mocks';
+import { ContentManagementPlugin } from './plugin';
+import { IRouter } from '@kbn/core/server';
+import { ProcedureName, procedureNames } from '../common';
+import type { Context, ProcedureDefinition } from './rpc';
+
+jest.mock('./core', () => ({
+  ...jest.requireActual('./core'),
+  Core: class {
+    setup() {
+      return {
+        contentRegistry: 'mockedContentRegistry',
+        api: {
+          register: jest.fn().mockReturnValue('mockedRegister'),
+          crud: jest.fn().mockReturnValue('mockedCrud'),
+          eventBus: {
+            emit: jest.fn().mockReturnValue('mockedEventBusEmit'),
+          },
+        },
+      };
+    }
+  },
+}));
+
+const mockGet = jest.fn().mockResolvedValue('getMocked');
+const mockCreate = jest.fn().mockResolvedValue('createMocked');
+const mockUpdate = jest.fn().mockResolvedValue('updateMocked');
+const mockDelete = jest.fn().mockResolvedValue('deleteMocked');
+const mockSearch = jest.fn().mockResolvedValue('searchMocked');
+
+jest.mock('./rpc/procedures/all_procedures', () => ({
+  procedures: {
+    get: {
+      fn: (...args: unknown[]) => mockGet(...args),
+      schemas: {
+        in: {
+          validate: () => undefined,
+        } as any,
+      },
+    },
+    create: {
+      fn: (...args: unknown[]) => mockCreate(...args),
+      schemas: {
+        in: {
+          validate: () => undefined,
+        } as any,
+      },
+    },
+    update: {
+      fn: (...args: unknown[]) => mockUpdate(...args),
+      schemas: {
+        in: {
+          validate: () => undefined,
+        } as any,
+      },
+    },
+    delete: {
+      fn: (...args: unknown[]) => mockDelete(...args),
+      schemas: {
+        in: {
+          validate: () => undefined,
+        } as any,
+      },
+    },
+    search: {
+      fn: (...args: unknown[]) => mockSearch(...args),
+      schemas: {
+        in: {
+          validate: () => undefined,
+        } as any,
+      },
+    },
+  } as { [key in ProcedureName]: ProcedureDefinition<Context, any, any> },
+}));
+
+const setup = () => {
+  const logger = loggingSystemMock.create();
+  const { http } = coreMock.createSetup();
+
+  const router: IRouter<any> = http.createRouter();
+  router.post = jest.fn();
+
+  const plugin = new ContentManagementPlugin({ logger });
+
+  return { plugin, http: { createRouter: () => router }, router };
+};
+
+describe('ContentManagementPlugin', () => {
```

Comment:
Do you think these tests should be end-to-end API tests or integration jest tests instead? 


It seems that as the regular module unit tests these tests aren't really that helpfull and could become a maintenance overhead compering to the value they'd bring.

Reply:
I disagree, I do think those tests are really helpfull. They immediately notify us if a server side handler has been declared for each of our procedures. They also test that all router routes are working for every procedures (declared in a single place in `common`).

With that confidence in place we can then fully unit test every procedure handler in isolation mocking their deps.

We will then have very few e2e tests needed which are _very_ slow to run and hard to debug.

Question: for api integration test we will first need a content to be registered. Do you know how we will be able to hook a "foo" content registration at Kibana bootstrap?

Reply:
> Question: for api integration test we will first need a content to be registered. Do you know how we will be able to hook a "foo" content registration at Kibana bootstrap?

One option is to run tests with a plugin that registers `foo`. Another option is running against `examples/` and having an example plugin that registers `foo`: can test both API and UI end-to-end.

Reply:
Mmmm not sure this will work for api integration. I really would like to avoid too much tests in e2e for our CM service. I much prefer API integration + jest comp integration that ensure we are calling the right routes with the right payloads.

We can discuss it at later sync 👍